lebedigital/demonstrator_scripts/computation_volume_content.py

```diff
@@ -0,0 +1,185 @@
+
+    # volume fractions
+    ## plasticizer
+    output['plasticizer_vol_fraction'] = input_dic['plasticizer_volume_content']/input_dic['density_plasticizer']
+    ## water
+    output['water_vol_fraction'] = (1 - input_dic['aggregates_volume_fraction'])*water_vol_fraction_to_binder\
+                                   - output['plasticizer_vol_fraction']
+    ## binder
+    binder_vol_fraction = (1 - input_dic['aggregates_volume_fraction'])*(1 - water_vol_fraction_to_binder)
+
+    output['sub_vol_fraction'] = binder_vol_fraction*input_dic['sc_volume_fraction']
+    output['cem_vol_fraction'] = binder_vol_fraction*(1-input_dic['sc_volume_fraction'])
+
+    # sanity check, the volume fractions add up to 1
+    assert 1 == pytest.approx((output['plasticizer_vol_fraction'] +
+                              output['water_vol_fraction'] +
+                              output['sub_vol_fraction'] +
+                              output['cem_vol_fraction'] +
+                              input_dic['aggregates_volume_fraction']).magnitude)
+
+    # computation of volume contents
+    output['cem_vol_content'] = output['cem_vol_fraction'] * input_dic['density_cem']
+    output['sub_vol_content'] = output['sub_vol_fraction'] * input_dic['density_sub']
+    output['water_vol_content'] = output['water_vol_fraction'] * input_dic['density_water']
+    output['aggregates_vol_content'] = input_dic['aggregates_volume_fraction'] * input_dic['density_aggregates']
+
+    # sanity check, the total volume adds up to 1
+    assert 1 == pytest.approx((input_dic['plasticizer_volume_content']/input_dic['density_plasticizer'] +
+                                        output['water_vol_content']/input_dic['density_water'] +
+                                        output['sub_vol_content']/input_dic['density_sub'] +
+                                        output['cem_vol_content']/input_dic['density_cem'] +
+                                        output['aggregates_vol_content']/input_dic['density_aggregates']).magnitude)
+
+    # paste density
+    output['density_paste'] = (output['cem_vol_content'] + output['sub_vol_content'] +
+                              output['water_vol_content'] + input_dic['plasticizer_volume_content']) /\
+                              (1-input_dic['aggregates_volume_fraction'])
+
+    return output
+
+
+
+def computation_ratios(input_dic):
+    """
+    This is the function to compute mass or volume ratios based on volume contents
+    It is basically the inverse of 'computation_volume_content'.
+
+    The input and output is wrapped by the python pint package: https://pint.readthedocs.io/
+    This requires units to be attached to the input values.
+
+    Parameters
+    ----------
+    input : dic / pint units required converted to matching units
+        required:
+            - 'water_vol_content'
+            - 'density_water'
+            - 'density_cem'
+            - 'cem_vol_content'
```

Comment:
we usually would have the cem_mass in a mix documentation

Reply:
this is the same